### PR TITLE
[VectorCombine] Add free concats to shuffleToIdentity.

### PR DESCRIPTION
**Pull request by davemgreen**
This is another relatively small adjustment to shuffleToIdentity, which has had a few knock-one effects to need a few more changes. It attempts to detect free concats, that will be legalized to multiple vector operations. For example if the lanes are '[a[0], a[1], b[0], b[1]]' and a and b are v2f64 under aarch64.

In order to do this:
 - isFreeConcat detects whether the input has piece-wise identities from multiple inputs that can become a concat.
 - A tree of concat shuffles is created to concatenate the input values into a single vector. This is a little different to most other inputs as there are created from multiple values that are being combined together, and we cannot rely on the Lane0 insert location always being valid.
 - The insert location is changed to the original location instead of updating per item, which ensure it is valid due to the order that we visit and create items.

### CODE REVIEW

**Comment by davemgreen** [comment]
Ping. I know there is talk of making other vectorizers perform re-vectorization, but I would like to get this in if we can. It helps fix a regression, Thanks.

**Comment by davemgreen** [comment]
Ping. Thanks

**Comment by RKSimon** [comment]
Sorry I thought this was related to the costmodel PR - I will take a look soon

File: llvm/lib/Transforms/Vectorize/VectorCombine.cpp [hunk 1]
```diff
@@ -1703,23 +1703,73 @@ generateInstLaneVectorFromOperand(ArrayRef<InstLane> Item, int Op) {
...
+static bool isFreeConcat(ArrayRef<InstLane> Item,
+                         const TargetTransformInfo &TTI) {
+  auto *Ty = cast<FixedVectorType>(Item.front().first->get()->getType());
+  unsigned NumElts = Ty->getNumElements();
+  if (Item.size() == NumElts || NumElts == 1 || Item.size() % NumElts != 0)
+    return false;
+
+  // Check that the concat is free, usually meaning that the type will be split
+  // during legalization.
+  SmallVector<int, 16> ConcatMask(Ty->getNumElements() * 2);
```
Comment (RKSimon) [inline llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1716]: `NumElts * 2` - but why hardcode to 2?

Comment (davemgreen) [reply llvm/lib/Transforms/Vectorize/VectorCombine.cpp:1716]: The shuffle costs are limited to 2 inputs, so it is measuring the cost of the first step in the tree it needs to create, and assuming that if that is free the whole thing will be free. Does that sound OK or do you think it should be measuring different levels?

**Review by davemgreen — COMMENTED** [review]
Thanks for taking a look. The cost-model issue I have not heard back from, so I believe / am hoping it was fixed by one of the backend patches.

File: llvm/lib/Transforms/Vectorize/VectorCombine.cpp (hunk 1 above)
Comment (RKSimon) [reply llvm/lib/Transforms/Vectorize/VectorCombine.cpp]: I think that should be OK.

**Review by RKSimon — APPROVED** [review]
LGTM

**Pull request merged**